lib/reports/rare_uncommitted.rb

```diff
@@ -0,0 +1,194 @@
+# frozen_string_literal: true
+
+require "cluster"
+require "services"
+Services.mongo!
+
+module Reports
+  # Finds clusters matching the criteria:
+  # has holdings,
+  # has no (active) commitments,
+  # has items with access:allow,
+  # has format: spm,
+  # has up to a specified no. of members holding it.
+  # The number of members can be counted as regular members, shared print members,
+  # or both (set independently).
+  # Usage:
+  #
+  # Find clusters held by up to 5 members:
+  # clusters = Reports::RareUncommitted.new.run(h: 5).to_a
+  #
+  # Find clusters held by up to 2 sp members:
+  # clusters = Reports::RareUncommitted.new.run(sph: 2).to_a
+  #
+  # Find clusters held by up to 5 members and up to 2 sp members:
+  # clusters = Reports::RareUncommitted.new.run(h: 5, sph: 2).to_a
+  class RareUncommitted
+    def initialize(memoize_orgs: true)
+      @memoize_orgs = memoize_orgs # memoize sp_organizations if true, turn off for testing
+      @query = {
+        "ht_items.0" => {"$exists": 1},
+        "holdings.0" => {"$exists": 1}
+      }
+    end
+
+    # All organizations that participate in the HT SP program.
+    def sp_organizations
+      if @memoize_orgs
+        @sp_organizations ||= Cluster.distinct("commitments.organization").compact
```

Comment:
Or this guarantees an accurate list and the upfront cost doesn't matter since this report won't be run too often.

Reply:
Presumably, commitments.organization is indexed? Since it's so quick I think it makes the most sense to use the actual holdings data rather than having potential mismatches with the ht_billing_members table.